tests/unit_tests/sparse_blas/source/sparse_spsv_usm.cpp

```diff
@@ -218,8 +218,11 @@ int test_spsv(sycl::device *dev, sparse_matrix_format_t format, intType m, doubl
                                 y_ref_host.data());
 
     // Compare the results of reference implementation and DPC++ implementation.
+    // Increase default relative error margin for tests that lead to large numeric values.
+    double abs_error_factor = 10;
+    double rel_error_factor = 1E5;
     ev_copy.wait_and_throw();
-    bool valid = check_equal_vector(y_host, y_ref_host);
+    bool valid = check_equal_vector(y_host, y_ref_host, abs_error_factor, rel_error_factor);
```

Comment:
Is this really necessary? How large are the numeric values? Looks like something may be amiss here. Does the matrix have all it's diagonal values non-zero? (Any triangular matrix with a zero value on its diagonal is singular, so the system is ill-posed.) We didn't need it before for the Intel oneMKL CPU/GPU backends, so why do we need this now?

Reply:
Some of the tests failed with incorrect results but were really close to the reference still. I wasn't able to reproduce recently however the issue was even worse with rocSPARSE. For rocSPARSE I had to set a fix seed (which is good practice to test operations very sensitive to the input values) and also change the range of values in the diagonal. With the current diagonal values in the range [0.1, 0.5] the output reference contains values between -1E11 and 1E11. Accurately comparing floating values in such a large range is not possible AFAIK. I have changed the diagonals to be in the range [10, 20] which gives a reference in the range [-20, 20]. With this change we don't need to modify the error bounds anymore, see https://github.com/oneapi-src/oneMKL/pull/527/commits/70cdbe09e5b0168821d5bdc15d43393cd98cf9ae.
For context these error factors are used with an heuristic [here](https://github.com/oneapi-src/oneMKL/blob/70cdbe09e5b0168821d5bdc15d43393cd98cf9ae/tests/unit_tests/sparse_blas/include/test_common.hpp#L525) and multiplied by epsilon [here](https://github.com/oneapi-src/oneMKL/blob/70cdbe09e5b0168821d5bdc15d43393cd98cf9ae/tests/unit_tests/sparse_blas/include/test_common.hpp#L493). Even using `rel_error_factor = 1E5` led to a reasonable relative error margin for large values in the end.
I was not surprised to see that different backends give different outputs for operations like spsv.